src/libmeasurement_kit/dns/system_resolver.hpp

```diff
@@ -0,0 +1,174 @@
+// information on the copying conditions.
+
+#include <measurement_kit/dns.hpp>
+
+#include <cassert>
+#include <future>
+
+#include <iostream>
+
+namespace mk {
+namespace dns {
+
+class ResolverContext {
+  public:
+    QueryClass dns_class;
+    QueryType dns_type;
+    std::string name;
+    Callback<Error, Var<Message>> cb;
+    Settings settings;
+    Var<Reactor> reactor;
+    Var<Logger> logger;
+
+    addrinfo hints;
+
+    Var<Message> message{new Message};
+
+    ResolverContext(QueryClass dns_c, QueryType dns_t, std::string n,
+                    Callback<Error, Var<Message>> c, Settings s, Var<Reactor> r,
+                    Var<Logger> l) {
+        dns_class = dns_c;
+        dns_type = dns_t;
+        name = n;
+        cb = c;
+        settings = s;
+        reactor = r;
+        logger = l;
+        memset(&hints, 0, sizeof(hints));
+    }
+};
+
+template <MK_MOCK(getaddrinfo), MK_MOCK(inet_ntop)>
+void resolve_async(ResolverContext *context) {
+    std::unique_ptr<ResolverContext> ctx(context);
+    Callback<Error, Var<Message>> callback = ctx->cb;
+    Var<Message> message = ctx->message;
+    addrinfo *servinfo_p = nullptr;
+
+    int error =
+        getaddrinfo(ctx->name.c_str(), nullptr, &ctx->hints, &servinfo_p);
+    std::unique_ptr<addrinfo, std::function<void(addrinfo *)>> servinfo(
+        servinfo_p, [](addrinfo *servinfo) {
+            if (servinfo != nullptr)
+                freeaddrinfo(servinfo);
+        });
+    if (error) {
+        ctx->logger->warn("getaddrinfo failed: %s", gai_strerror(error));
+        Error resolver_error;
+        switch (error) {
+        case EAI_AGAIN:
+            resolver_error = TemporaryFailure();
+            break;
+        case EAI_BADFLAGS:
+            resolver_error = InvalidFlagsValue();
+            break;
+        case EAI_BADHINTS:
+            resolver_error = InvalidHintsValue();
+            break;
+        case EAI_FAIL:
+            resolver_error = NonRecoverableFailure();
+            break;
+        case EAI_FAMILY:
+            resolver_error = NotSupportedAIFamily();
+            break;
+        case EAI_MEMORY:
+            resolver_error = MemoryAllocationFailure();
+            break;
+        case EAI_NONAME:
+            resolver_error = HostOrServiceNotProvidedOrNotKnown();
+            break;
+        case EAI_OVERFLOW:
+            resolver_error = ArgumentBufferOverflow();
+            break;
+        case EAI_PROTOCOL:
+            resolver_error = UnknownResolvedProtocol();
+            break;
+        case EAI_SERVICE:
+            resolver_error = NotSupportedServname();
+            break;
+        case EAI_SOCKTYPE:
+            resolver_error = NotSupportedAISocktype();
+            break;
+        default:
+            resolver_error = ResolverError();
+        }
+        ctx->reactor->call_soon([=]() { callback(resolver_error, nullptr); });
+        return;
+    }
+    assert(servinfo != nullptr);
```

Comment:
Since `serveinfo` is a smart pointer it could also be written as `assert(!!servinfo)`